### ISSUE TITLE
Different behavior between TruffleRuby and CRuby

### ISSUE THREAD

**Issue by Maumagnaguagno** [oracle/truffleruby#2658]
The following code has different behavior in TruffleRuby 22.1.0 and CRuby 3.1.2.
```Ruby
a = [1,2,3]
h = {}
until a.empty?
  p a
  h[a.shift] ||= false
end
p h
```
See outputs below:
```Ruby
# CRuby
[1, 2, 3]
[2, 3]
[3]
{1=>false, 2=>false, 3=>false}

# TruffleRuby
[1, 2, 3]
[3]
{2=>false, nil=>false}
```

To my surprise, replacing ``||=`` with ``=`` makes both Rubies output the same.
Forgive me if this is already fixed or if there is a related open issue, my search resulted in nothing similar.

**Comment by eregon** [comment]
This is because the `a.shift` is evaluated twice by TruffleRuby, but only once by CRuby.
We should fix that..

This variant has no issue:
```ruby
a = [1,2,3]
h = {}
until a.empty?
  p a
  e = a.shift
  h[e] ||= false
end
p h
```

**Comment by aardvark179** [comment]
Our translation of operator assignment nodes seems half correct. We correctly store the receiver as a local variable, but we do not do this correctly with any arguments passed to that receiver.

**Comment by aardvark179** [comment]
The fix for this has now been merged into master.
